File: app/api/src/Astra.Api/Llm/Archetypes/java-spring/canonical-php-catalog-service/src/main/java/com/acme/catalog/PriceFetcher.java
// SPDX-Spec: php/fetch_price.php (signed)
// SPDX-Archetype: canonical-php-catalog-service
package com.acme.catalog;

import java.math.BigDecimal;
import java.util.Optional;

/**
 * Java 21 projection of PHP {@code fetch_price.php}. The PHP source used
 * {@code @file_get_contents} (suppresses the warning) and {@code json_decode}
 * (returns null on failure), so a network/parse failure was silently swallowed
 * and the caller received {@code 0.0} — indistinguishable from a real zero price.
 *
 * <p>EH-1: the migration makes the failure EXPLICIT. Two flavours are provided so
 * the SME can choose the contract on sign-off:
 * <ul>
 *   <li>{@link #fetchOrThrow(String)} — throw {@link PriceUnavailableException}
 *       (recommended when a missing price must halt checkout).</li>
 *   <li>{@link #fetchOrDefault(String, BigDecimal)} — fall back to an EXPLICIT
 *       caller-supplied default (never a silent 0.0).</li>
 * </ul>
 */
@TargetMapping(value = "@Service", phpConstruct = "fetch_price.php")
public final class PriceFetcher {

    private final RemotePricePort port;

    public PriceFetcher(RemotePricePort port) {
        this.port = port;
    }

    /** Fetch the price or throw — the failure is never swallowed. */
    @SpecClaim("EH-1")
    public BigDecimal fetchOrThrow(String url) {
        return port.fetch(url).orElseThrow(() -> new PriceUnavailableException(url));
    }

    /** Fetch the price or fall back to an EXPLICIT default (not a silent 0.0). */
    @SpecClaim("EH-1")
    public BigDecimal fetchOrDefault(String url, BigDecimal fallback) {
        if (fallback == null) {
            throw new IllegalArgumentException("fallback is required (no silent 0.0)");
        }
        Optional<BigDecimal> price = port.fetch(url);
        return price.orElse(fallback);
    }
}

File: app/api/src/Astra.Api/Llm/Archetypes/dotnet8/canonical-php-catalog-service/src/PriceFetcher.cs
// SPDX-Spec: php/fetch_price.php (signed)
// SPDX-Archetype: canonical-php-catalog-service (dotnet8)
namespace Acme.Catalog;

/// <summary>
/// Port for the PHP <c>@file_get_contents($url)</c> + <c>json_decode</c> lookup.
/// A null result models the PHP failure path (false → null) so the caller must
/// handle absence EXPLICITLY rather than via the silent <c>@</c> operator.
/// </summary>
[TargetMapping("HttpClient price gateway", PhpConstruct = "@file_get_contents + json_decode")]
public interface IRemotePrice
{
    /// <summary>The remote price, or null if unreachable/unparseable.</summary>
    [SpecClaim("SE-1")]
    decimal? Fetch(string url);
}

/// <summary>Typed exception replacing PHP's silent @-suppressed failure (EH-1).</summary>
public sealed class PriceUnavailableException : Exception
{
    public PriceUnavailableException(string url)
        : base($"Remote price unavailable for: {url}") { }
}

/// <summary>
/// C# projection of PHP fetch_price.php. PHP's <c>@</c> + false-return swallowed
/// failures and returned 0.0. EH-1: the migration surfaces failure EXPLICITLY —
/// throw, or fall back to a caller-supplied default, never a silent 0.
/// </summary>
[TargetMapping("service class", PhpConstruct = "fetch_price.php")]
public sealed class PriceFetcher
{
    private readonly IRemotePrice _port;

    public PriceFetcher(IRemotePrice port) => _port = port;

    /// <summary>Fetch the price or throw — never swallowed.</summary>
    [SpecClaim("EH-1")]
    public decimal FetchOrThrow(string url)
        => _port.Fetch(url) ?? throw new PriceUnavailableException(url);

    /// <summary>Fetch the price or fall back to an EXPLICIT default (not a silent 0).</summary>
    [SpecClaim("EH-1")]
    public decimal FetchOrDefault(string url, decimal fallback)
        => _port.Fetch(url) ?? fallback;
}
